source/s3_auto_ranged_put.c

```diff
@@ -684,244 +743,545 @@ static int s_skip_parts_from_stream(
         num_parts_read_from_stream,
         skip_until_part_number);
 
-    int return_status = AWS_OP_SUCCESS;
-    for (uint32_t part_index = num_parts_read_from_stream; part_index < skip_until_part_number; ++part_index) {
+    /* Kick off async work loop */
+    s_skip_parts_from_stream_loop(NULL, skip_ctx);
+
+    return skip_future;
+}
+
+/* Async work loop for reading over skipped parts.
+ * In a loop, try to read each part. If the read completes immediately then keep looping.
+ * Otherwise, register this same function as the completion callback and bail out,
+ * we'll resume the loop when the callback fires.
+ *
+ * It would have been simpler to always set a completion callback,
+ * but this risks the call stack growing very large (if there are many parts
+ * to skip and the reads all complete immediately). */
+static void s_skip_parts_from_stream_loop(struct aws_future *read_part_future, void *user_data) {
+    struct aws_s3_skip_parts_from_stream_async_ctx *skip_ctx = user_data;
+    struct aws_s3_meta_request *meta_request = skip_ctx->meta_request;
+    struct aws_s3_auto_ranged_put *auto_ranged_put = meta_request->impl;
+    struct aws_byte_buf *temp_body_buf = &skip_ctx->temp_body_buf;
+
+    int error_code = AWS_ERROR_SUCCESS;
 
-        size_t request_body_size = s_compute_request_body_size(meta_request, part_index + 1);
+    for (; skip_ctx->part_index < skip_ctx->skip_until_part_number; ++skip_ctx->part_index) {
 
-        if (temp_body_buf.capacity != request_body_size) {
-            // reinit with correct size
-            aws_byte_buf_clean_up(&temp_body_buf);
-            if (aws_byte_buf_init(&temp_body_buf, meta_request->allocator, request_body_size)) {
-                return AWS_OP_ERR;
+        /* unless we just re-entered loop due to async read completion, kick off a new async read */
+        if (read_part_future == NULL) {
+            size_t request_body_size = s_compute_request_body_size(meta_request, skip_ctx->part_index + 1);
+
+            if (temp_body_buf->capacity != request_body_size) {
+                // reinit with correct size
+                aws_byte_buf_clean_up(temp_body_buf);
+                aws_byte_buf_init(temp_body_buf, meta_request->allocator, request_body_size);
+            } else {
+                // reuse buffer
+                aws_byte_buf_reset(temp_body_buf, false);
+            }
+
+            read_part_future = aws_s3_meta_request_read_body(skip_ctx->meta_request, &skip_ctx->temp_body_buf);
+
+            /* the read may or may not complete immediately */
+            if (aws_future_register_callback_if_not_done(read_part_future, s_skip_parts_from_stream_loop, skip_ctx)) {
```

Comment:
This is the worst-named function in the aws_future API
I am 1000% open to suggestions

I used to call it `aws_future_is_done_else_register_callback()` and it would return whether the future was done. But then everywhere I used it, I wanted to early-out if it was NOT done, and putting a `!`  or `== false` made it even harder to read, so I flipped what made it return true, and renamed it and 🤢